src/Autonomous.cpp

```diff
@@ -243,23 +264,99 @@ void Autonomous::autonomyIter()
 			// caused by filtering
-			driveTarget.topRows(2) = landmarkFilter.get(landmarkMapSpace).topRows(2);
-			// clear the filter so the second post is removed from it
-			landmarkFilter.reset();
-		}
-		// If we are in a search pattern, set the drive target to the next search point
-		else if (state == NavState::SEARCH_PATTERN)
-		{
-			driveTarget = search_target;
+			point_t post_2 = gate_filter.get(landmarkMapSpace);
+
+			// we haven't seen the first post yet, use the second post as the drive target
+			if (landmarkFilter.getSize() == 0)
+			{
+				driveTarget.topRows(2) = post_2.topRows(2);
+			}
+			// we have already seen the first post
+			else
+			{
+				point_t post_1 = landmarkFilter.get();
+				point_t center = (post_1 + post_2) / 2;
+
+				// Use the center as the drive target to avoid being close to one post but far from the other
+				driveTarget.topRows(2) = center.topRows(2);
+
+				// True if we are close enough to have sufficiently accurate landmark measurements
+				bool inRange = std::isnan(gate_targets.first(2)) && dist(pose, center, 0.0) < 8.0;
+				// True if we are within refining distance
+				bool shouldRefine = !std::isnan(gate_targets.first(2)) && std::isnan(gate_targets.second(2)) && dist(pose, gate_targets.first, 0.0) < 1.5;
+
+				// we have not set gate targets yet or have potentially inaccurate gate landmark measurements
+				if (inRange || shouldRefine)
+				{
+					point_t offset{(post_1(1) - post_2(1)) / 2, (post_2(0) - post_1(0)) / 2, 0};
+					pose_t goal_1 = center + 2 * offset;
+					pose_t goal_2 = center - 2 * offset;
+
+					// Choose the closest goal to be the first target
+					if (dist(pose, goal_1, 0) < dist(pose, goal_2, 0))
+					{
+						gate_targets = std::make_pair(goal_1, goal_2);
+					}
+					else
+					{
+						gate_targets = std::make_pair(goal_2, goal_1);
+					}
+
+					// Angle between targets
+					double angle = std::atan2(gate_targets.second(1) - gate_targets.first(1), gate_targets.second(0) - gate_targets.first(0));
+					// Angle between current post and first target
+					gate_targets.first(2) = std::atan2(gate_targets.first(1) - pose(1), gate_targets.first(0) - pose(0));
+					// Mark targets as unrefined if not refining
+					gate_targets.second(2) = shouldRefine ? angle : NAN;
+
+					// Replan
+					plan_cost = INFINITE_COST;
+
+					// Align to target to improve planning
+					state = NavState::GATE_ALIGN;
+				}
+
+				// we have already seen the first post and have set gate targets
+				if (!std::isnan(gate_targets.first(2)))
+				{
+					// Update gate targets
+					if (!std::isinf(gate_targets.first(2)) &&
+						dist(pose, gate_targets.first, 0.0) < 0.2)
+					{
+						// Arrived at first gate target, mark as complete
+						gate_targets.first = pose_t(INFINITY, INFINITY, INFINITY);
+
+						// Update target angle of second target to account for inaccuracies
+						gate_targets.second(2) = std::atan2(gate_targets.second(1) - pose(1), gate_targets.second(0) - pose(0));
```

Comment:
Again, I think we should handle this just by constantly recomputing both gate targets (earlier in the control loop, where I left that other comment).